src/vulkan-renderer/imgui/imgui.cpp

```diff
@@ -0,0 +1,616 @@
+    pipelineCreateInfo.pColorBlendState = &pipelineColorBlendStateCreateInfo;
+    pipelineCreateInfo.pMultisampleState = &pipelineMultisampleStateCreateInfo;
+    pipelineCreateInfo.pViewportState = &pipelineViewportStateCreateInfo;
+    pipelineCreateInfo.pDepthStencilState = &pipelineDepthStencilStateCreateInfo;
+    pipelineCreateInfo.pDynamicState = &pipelineDynamicStateCreateInfo;
+    pipelineCreateInfo.stageCount = static_cast<uint32_t>(shaders.size());
+    pipelineCreateInfo.pStages = shaders.data();
+    pipelineCreateInfo.subpass = subpass;
+
+    VkVertexInputBindingDescription vInputBindDescription{};
+
+    vInputBindDescription.binding = 0;
+    vInputBindDescription.stride = sizeof(ImDrawVert);
+    vInputBindDescription.inputRate = VK_VERTEX_INPUT_RATE_VERTEX;
+
+    // TODO() Clean up!
+    std::vector<VkVertexInputBindingDescription> vertexInputBindings = {vInputBindDescription};
+
+    std::vector<VkVertexInputAttributeDescription> vertexInputAttributes(3);
+
+    // Location 0: Position
+    vertexInputAttributes[0].binding = 0;
+    vertexInputAttributes[0].location = 0;
+    vertexInputAttributes[0].format = VK_FORMAT_R32G32_SFLOAT;
+    vertexInputAttributes[0].offset = offsetof(ImDrawVert, pos);
+
+    // Location 1: UV
+    vertexInputAttributes[1].binding = 0;
+    vertexInputAttributes[1].location = 1;
+    vertexInputAttributes[1].format = VK_FORMAT_R32G32_SFLOAT;
+    vertexInputAttributes[1].offset = offsetof(ImDrawVert, uv);
+
+    // Location 0: Color
+    vertexInputAttributes[2].binding = 0;
+    vertexInputAttributes[2].location = 2;
+    vertexInputAttributes[2].format = VK_FORMAT_R8G8B8A8_UNORM;
+    vertexInputAttributes[2].offset = offsetof(ImDrawVert, col);
+
+    VkPipelineVertexInputStateCreateInfo pipelineVertexInputStateCreateInfo{};
+
+    pipelineVertexInputStateCreateInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO;
+    pipelineVertexInputStateCreateInfo.vertexBindingDescriptionCount =
+        static_cast<uint32_t>(vertexInputBindings.size());
+    pipelineVertexInputStateCreateInfo.pVertexBindingDescriptions = vertexInputBindings.data();
+    pipelineVertexInputStateCreateInfo.vertexAttributeDescriptionCount =
+        static_cast<uint32_t>(vertexInputAttributes.size());
+    pipelineVertexInputStateCreateInfo.pVertexAttributeDescriptions = vertexInputAttributes.data();
+
+    pipelineCreateInfo.pVertexInputState = &pipelineVertexInputStateCreateInfo;
+
+    spdlog::debug("Creating ImGUI graphics pipeline");
+
+    m_pipeline = std::make_unique<wrapper::GraphicsPipeline>(
+        m_device.device(), m_pipeline_layout, m_renderpass->get(), shaders, vertexInputBindings, vertexInputAttributes,
+        m_swapchain.extent().width, m_swapchain.extent().height, "ImGUI");
+
+    update();
+    prepare_drawing();
+}
+
+ImGUIOverlay::~ImGUIOverlay() {
+    free_resources();
+}
+
+bool ImGUIOverlay::update() {
```

Comment:
We should focus our optimization efforts here, (I think) it's currently the leading cause of the huge FPS drop.

Reply:
I agree.

Reply:
Let's discuss how to proceed with the update method next.